### PR TITLE
Add SNS process, improve error messages, refactor

### PR DESCRIPTION
**Pull request by hrbigelow**
This refactors the dispatch logic previously in builder.py and places all model-specific code in transition_matrices.py.
The advantage here is that adding a new model now entails only edits to transition_matrices.py.

Also, please see the even_ones implementation change - I believe this now matches the figure in the paper.

### CODE REVIEW

File: simplexity/generative_processes/transition_matrices.py [hunk 1]
```diff
@@ -47,10 +55,36 @@ def even_ones(p: float) -> jax.Array:
             ],
             [
                 [0, p],
-                [1, 0],
+                [0, 1],
```
Comment (ealt) [inline simplexity/generative_processes/transition_matrices.py:58]: Doesn't this imply P(obs=1, next_state=1 | curr_state=1) = 1, when what we want is P(obs=1, next_state=0 | curr_state=1) = 1

File: simplexity/generative_processes/transition_matrices.py [hunk 2]
```diff
@@ -169,23 +203,29 @@ def _intermediate_matrix(val):
...
-    transition_matrices = transition_matrices.at[1, s["S"], s["1"]].set(1 - pR1)
-    transition_matrices = transition_matrices.at[0, s["0"], s["F"]].set(pR2)
-    transition_matrices = transition_matrices.at[1, s["0"], s["T"]].set(1 - pR2)
-    transition_matrices = transition_matrices.at[0, s["1"], s["T"]].set(pR2)
-    transition_matrices = transition_matrices.at[1, s["1"], s["F"]].set(1 - pR2)
-    transition_matrices = transition_matrices.at[1, s["T"], s["S"]].set(1.0)
-    transition_matrices = transition_matrices.at[0, s["F"], s["S"]].set(1.0)
+    """Random-random Exclusive-Or Process."""
+    p, q = pR1, 1 - pR1
+    x, y = pR2, 1 - pR2
```
Comment (ealt) [inline simplexity/generative_processes/transition_matrices.py:208]: The original version makes is easier to understand what all the non-zero terms of the transition matrices represent, and given that instantiating the matrices is cheap and generally happens once per run I would prefer not to trade off more intuitive human readability for marginal improvements in simplicity and performance.

File: simplexity/generative_processes/transition_matrices.py [hunk 3]
```diff
@@ -1,6 +1,13 @@
+from enum import Enum
+
 import jax
 import jax.numpy as jnp
 
+"""
+Processes are defined by a tensor T_kij parameterizing P(X_k, S_j | \\hat{S}_i).
+X, S, \\hat{S} are respectively: current observed, current state, previous state.
```
Comment (ealt) [inline simplexity/generative_processes/transition_matrices.py:8]: why use hat for previous but not current? And why use i,j,k indexes above and drop them here?

File: simplexity/generative_processes/transition_matrices.py [hunk 4]
```diff
@@ -47,10 +55,36 @@ def even_ones(p: float) -> jax.Array:
...
+        ]
+    )
+
+
+def sns(p: float, q: float):
+    """Creates a transition matrix for the Simple Nonunifilar Source Process.
+
+    Defined in https://arxiv.org/pdf/1702.08565 Fig 2.
+    """
+    T_kij = jnp.array(
```
[the remark this answers is not in the document]
Comment (hrbigelow) [reply simplexity/generative_processes/transition_matrices.py:69]: sg

File: simplexity/generative_processes/transition_matrices.py [hunk 5]
```diff
@@ -47,10 +55,36 @@ def even_ones(p: float) -> jax.Array:
...
+                [q, 0],
             ],
         ]
     )
+    return T_kij
+
+
+def coin(p: float):
+    """Create a transition matrix for a simple coin-flip Process."""
+    T_kij = jnp.array([[[p]], [[1 - p]]])
```
[the remark this answers is not in the document]
Comment (hrbigelow) [reply simplexity/generative_processes/transition_matrices.py:86]: sg

File: simplexity/generative_processes/transition_matrices.py (hunk 1 above)
Comment (hrbigelow) [reply simplexity/generative_processes/transition_matrices.py:58]: image

```python
        [
            [
                [q, 0],
                [0, 0],
            ],
            [
                [0, p],
                [0, 1],
            ],
        ]
 ```

File: simplexity/generative_processes/transition_matrices.py (hunk 2 above)
Comment (hrbigelow) [reply simplexity/generative_processes/transition_matrices.py:208]: okay sounds good

File: simplexity/generative_processes/transition_matrices.py (hunk 1 above)
Comment (hrbigelow) [reply simplexity/generative_processes/transition_matrices.py:58]: Yes, I made a mistake in interpreting the figure

File: simplexity/generative_processes/transition_matrices.py (hunk 3 above)
Comment (hrbigelow) [reply simplexity/generative_processes/transition_matrices.py:8]: I updated comment

**Review by ealt — APPROVED** [review]
